ascpu.man
.TH ascpu 1 "10 October 2005" "Version 1.11"
.UC
.SH NAME
\fBascpu\fP \- the AfterStep CPU load monitor
.SH SYNOPSIS
.B ascpu
[-h] [-H] [-V]
        [-iconic] [-withdrawn] [-standout]
        [-position \fI[+|-]x[+|-]y\fP]
        [-dev \fIdevice\fP]
        [-cpu \fInumber\fP]
        [-u \fIupdate rate\fP]
        [-samples \fInumber\fP]
        [-history \fInumber\fP]
        [-nonice]
        [-display \fIdisplay\fP]
        [-title \fIname\fP]
        [-exe \fIcommand\fP]
        [-bg \fIcolor\fP]
        [-fg \fIcolor\fP]
        [-sys \fIcolor\fP]
        [-nice \fIcolor\fP]
        [-user \fIcolor\fP]
        [-idle \fIcolor\fP]
.SH DESCRIPTION
The \fBascpu\fP is a X11 application that acts as
an a CPU load monitor for computers running Linux,
FreeBSD or HP-UX.
The \fBascpu\fP provides a distinctive Afterstep
window manager look and feel and features multiple
options to allow the customization.
.PP
The right (big) area displays the running history
of the CPU load. With the default update period of
1 second it shows the current CPU load. It can be
made to show average CPU load in bigger steps.
.PP
On the left side you have an indicator that shows
the same values taken average over a number of samples.
The default number of samples is 60. This means
with the default setting of update rate
you see the average CPU load during the last minute.
If you change it to 1 second you'll see the current
CPU load there.
.PP
The load indicators are divided (by using different colors) 
into three parts: system CPU time (bright), nice CPU
time (medium), and the user CPU time (dark).
The consumed CPU time is displayed in percents of
total (plus idle time that is) CPU time passed.

.SH CONFIGURATION OPTIONS
.IP "-h or -H"
.RS
prints a short description and usage message.
.RE
.IP "-V"
.RS
Version control. Prints out the version of the program.
.RE
.IP "-position [+|-]x[+|-]y"
.RS
Displays the window at the specified location
on the screen. This works as standard X Windows
geometry option.
.RE
.IP "-iconic"
.RS
Starts the \fBascpu\fP application in the iconized mode.
The icon has the same appearance with the main
window.
When this option is given, the -position refers to
the position of the icon window as well as the main window.
.RE
.IP "-withdrawn"
.RS
Starts the \fBascpu\fP application in the withdrawn mode.
This option is necessary to be able to dock the
\fBascpu\fP in the WindowMaker dock.
When this option is given, the -position refers to
the position of the icon window as well as the main window.
\fBascpu\fP ignores the -iconic option when started
in withdrawn mode.
.RE
.IP "-standout"
.RS
This changes the appearance of the \fBascpu\fP display from
"pushed-in" to "popped-out". The first is the default look.
According to some the second looks much better in WindowMaker's dock.
Give it a try in any case.
.RE
.IP "-dev <device>"
.RS
Using this option you can specify the location of the file
that serves the CPU statistics information. The default
location of the this device on Linux systems is /proc/stat.
.RE
.IP "-cpu <number>"
.RS
Without this parameter the \fBascpu\fP displays statistics
for the overall system usage on both single-CPU and SMP
machines. On SMP machines, however, it is possible to obtain
statistics per CPU. Using this parameter you give the CPU
number (counted from zero) that you wish to watch.

If you give a CPU number which is bigger than any you actually
have on your machine (or any CPU number on a single-CPU
machine) \fBascpu\fP will not show anything in its window
and start complaining about "invalid character while reading
/proc/stat" on the stnadard output.

This works only on Linux machines at the moment (I do not
have a FreeBSD machine myself).
.RE
.IP "-u <update rate>"
.RS
Changes the polling rate for accessing the CPU
statistics data. The "-samples" and "-history" are
specified in units of this polling rate.
The update rate is specified in seconds. The value
of 1 is the minimum. Default value is 1 second.

This is useful to change when you use value of 1 for 
neither -samples nor -history. For example, if you want
to use "-samples 2 -history 60" (the bar on the left shows
average load over two seconds and the running history
shows load update every one minute) you should better
use "-u 2 -samples 1 -history 30" to decrease the load
on the CPU produced by the CPU monitor itself.
.RE
.IP "-samples <number>"
.RS
The number of samples that should be taken for the calculations
of the average load on the CPU (the bar on the left). The default 
value is 60.  The minimum number of samples that can be specified 
is 1. At the sample rate of 1 you turn it into the current load
indicator. While it it used as an average load indicator
(sample rate > 1) the average value is calculated statistically
over the last "<number>" of samples.
.RE
.IP "-history <number>"
.RS
The number of samples that go into one value displayed
in the running history window. The CPU load is sampled for
the specified number of cycles and the value is displayed.
The default value is 1.
.RE
.IP "-nonice"
.RS
Force \fBascpu\fP to show the nice CPU time as idle. This
may become handy if you run some task "niced" in the 
background and you do not want to see the CPU indicator to
show 100% load all the time. If this switch is specified the
nice CPU time is simply added to the idle CPU time and you
see CPU load only if user/system CPU cycles are consumed.
.RE
.IP "-display <name>"
.RS
The name of the display to start the window in.
It works just as X Windows display option.
.RE
.IP "-title <name>"
.RS
Set the window title and the icon title to the specified name.
.RE
.IP "-exe <command>"
.RS
Execute the given command when the applet is clicked with the mouse.
.RE
.IP "-fg <color>"
.RS
The color for the "foreground". This color is used to draw
the system CPU time. It is darkened by 1.2 to draw the nice
CPU time. It is darkened by 1.4 to draw the user CPU time.
See options -sys, -nice, -user to use different colors.
.RE
.IP "-bg <color>"
.RS
The color for the background of the meter. It is used to draw 
the idle CPU time and it is lightened a bit and darkened a bit
to create the 3D appearance of the ground plate.
See -idle option to draw idle time in a different color.
.RE
.IP "-sys <color>"
.RS
The color to use for the system CPU time. It overrides the
color specified with the -fg option.
.RE
.IP "-nice <color>"
.RS
The color to use for the nice CPU time. It overrides the
color specified with the -fg option.
.RE
.IP "-user <color>"
.RS
The color to use for the user CPU time. It overrides the
color specified with the -fg option.
.RE
.IP "-idle <color>"
.RS
The color to use for the idle CPU time. It overrides the
color specified with the -bg option.
.RE
.SH INVOCATION
\fBascpu\fP can be called in different ways.  The most 
common invocation is the command line:
.nf

	user@host[1]% ascpu -history 15 -samples 1 -fg "#d04040" &

.fi
Another way to call \fBascpu\fP is from the window manager:
.nf

	*Wharf "ascpu" nil Swallow "ascpu" ascpu -u 2 -samples 15 &

.fi
This line, when placed in the \fIwharf\fP file in the users Afterstep
configuration directory will cause \fBascpu\fP to be a
button on the \fBWharf (1)\fP button bar under the \fBafterstep (1)\fP
window manager.

If you have an SMP machine with two CPUs you can watch both
CPUs separately and the overall statistics at the same time:
.nf
	*Wharf "ascpu" nil Swallow "ascpu" ascpu &
	*Wharf "ascpu-0" nil Swallow "ascpu-0" ascpu -cpu 0 -title ascpu-0 &
	*Wharf "ascpu-1" nil Swallow "ascpu-1" ascpu -cpu 1 -title ascpu-1 &
.fi
This will result in three \fBascpu\fP windows in your Wharf
each showing the respective statistics.

If you run WindowMaker then you should use the "-withdrawn"
option:
.nf

	user@host[1]% ascpu -withdrawn -standout &

.fi
and then drag the icon to the dock.
.SH BUGS
My programs do not have bugs, they just develop random
features ;-)
.PP
Well, there are limitations. All the strings for the
color names, display name, and the geometry have the
length limit of 50 characters (terminating zero included). 
The string copying routine cuts the names that are longer.
The limit on the statistics file name is 256 characters.
.PP
I had to limit the number of the CPUs in the SMP system
you are using (lucky you :). Currently up to the CPU number
15 can be shown (that's 16 of them). If you want more you
will have to edit the header file state.h and recompile.
.PP
The SMP support is not available for FreeBSD at the moment.
Any volunteers to submit the necessary info and lend a hand
with development and testing?
.SH FILES
.I /proc/stat
.SH "SEE ALSO"
.BR top (1), proc (5)
.SH COPYRIGHTS
Copyright (c) 1998-2005  Albert Dorofeev <albert@tigr.net>

Distributed under GNU General Public License v2 ; 
see LICENSE file for more informations.
.SH AUTHORS
Albert Dorofeev <albert@tigr.net>

See README file for credits.
.sp


CHANGES
History:

0.1	some time in August 1998

	Original version, shows the same as asload.

0.2	20 September 1998

	The version that works and shows the average.
	I am looking for a better way to calculate
	average values so the 1.0 must wait.

1.0	22 September 1998

	Finally I got it all nice working. I could not
	figure an algorithm that would allow to adjust
	the number of samples without dynamic allocation
	of memory though. Probably does not exist ;-)

1.1	03 November 1998

	Oh, well, the whole thing rewritten almost
	from scratch again. It all started with the
	message from Cory Visi:

I have suggestion for your cool little Wharf app ascpu. I think that the CPU
meter on the side should have an interval independent of the graph display.
The CPU meter on the left is much for useful in real time (little or no
interval) while the graph is much more useful with a longer interval (like
10 seconds). So even if you hardcode a 0 or 1 interval for the CPU meter, I
think that would make it a lot nicer. I use an interval of 10 seconds, so
the CPU bar is pretty useless :)

	So, it runs in both conventional (as I dreamed it
	up) and the opposite (as Cory Visi dreamed it up)
	modes. The intervals are independent :)

1.2	02 January 1999

	Fixed the "-position" option to work with the 
	negative coordinates.

1.3	07 February 1999

	Thanks to Liang Tai-hwa <avatar@www.mmlab.cse.yzu.edu.tw>
	the ascpu now runs on FreeBSD systems as well.

1.4	12 April 1999

	Thanks to Stewart Gebbie <stewart@global.co.za> for fixing
	this embarrasing bug for me. He writes:
	"I use ascpu but noticed that the -user option didn't
	seem to work. Apon closer inspection I noticed that
	the code to check for the -user option was after
	the -u check and thus the 'if...else' statments
	picked up the -user as a -u.
	Attached is a patch for ascpu.c which fixes things."

1.5	28 April 1999

	Thanks to Evgeny Stambulchik <fnevgeny@plasma-gate.weizmann.ac.il>
	for fixing an annoying bug with "troughs" in the display.
	Apparently it is possible that the number of CPU cycles
	decreases by 1 under heavy load... Weird.

	Evgeny also helped me to set up the SMP support which
	goes out with this version.

	I added the -title option to be able to swallow a number
	of ascpu instances into the Wharf on an SMP machine.

1.6	06 May 1999

	Thanks to ii1 <ii1@grant.kharkov.ua> who sent in a patch
	ascpu can execute commands when clicked on now.

1.7	27 July 1999

	Stasinos Konstantopoulos <konstant@let.rug.nl> ported
	ascpu to HP-UX and kindly sent in the patch.

1.8	16 December 1999

	Changed the internal loop to react to the absolute
	value of the difference in time for the update.
	Suggested by Nicola Bernardini <nicb@axnet.it>

1.9	09 March 2000

	Fixes the security problem on the systems where ascpu
	must run with some kmem or root privileges (SUID)
	to access the system information (FreeBSD and now AIX).
	It was possible to execute arbitrary commands using
	the privilege. Now ascpu forks and the child drops
	privileges to execute the command.

	Kirill Bogachev <bogachev@mech.math.msu.su> ported
	ascpu to AIX on PowerPC:
	"I have ported it to AIX running on PowerPC. The port has been tested
	on AIX 4.1 but should works on AIX > 3.1.
	...
	In AIX "ascpu" should be installed setuid root to get access to
	kernel memory."

1.10	17 January 2004

	Patch by Mathieu Chouquet-Stringer <mathieu@newview.com>
	for Linux Kernel >= 2.5
	Patch by Jeff Brown <jabrown@caida.org> for script
	execution from "withdrawn" state.

1.11	10 October 2005

	Thanks to Kevin Chan-Henry <kch1@dsl.pipex.com> for
	pointing out that the Linux kernels after 2.6.10
	introduce an additional field in /proc/stat.
	If you never used the "-cpu" switch you will not be
	affected. But with multiple CPUs ascpu breaks.
	So this version should fix that.



ascpu.c
/*
 * ascpu is the CPU statistics monitor utility for X Windows
 * Copyright (c) 1998-2005  Albert Dorofeev <albert@tigr.net>
 * For the updates see http://www.tigr.net/
 *
 * This software is distributed under GPL. For details see LICENSE file.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "safecopy.h"

#include "ascpu_x.h"
#include "state.h"

extern struct ascpu_state state;

/*
 * default check and update intervals in microseconds
 *      x11 events - every 1/100 th of a second (in mks)
 *      CPU status - every second (in sec)
 *
 *      Collect 60 samples for the average load
 *      Collect 1 samples for the history load window
 */
#define X11_INTERVAL    10000L
#define CHK_INTERVAL    1
#define AVG_SAMPLES	60
#define HIST_SAMPLES	1

#define FOR_CLICK 1

int     withdrawn = 0;
int     iconic = 0;
int     pushed_in = 1;
char    display_name[50];
char    mainGeometry[50];
char    window_name[50];

#ifdef FOR_CLICK
char	Command[255]="";
#endif

/* colors */
int sys_color_defined = 0;
char sys_color[50];
int nice_color_defined = 0;
char nice_color[50];
int user_color_defined = 0;
char user_color[50];
int idle_color_defined = 0;
char idle_color[50];

void defaults()
{
	state.update_interval = CHK_INTERVAL;
	state.avg_samples = AVG_SAMPLES;
	state.hist_samples = HIST_SAMPLES;
	state.no_nice = 0;
	state.cpu_number = -1;
	safecopy(state.proc_stat_filename, PROC_STAT, 256);
	withdrawn = 0;
	iconic = 0; 
	pushed_in = 1;
	safecopy(window_name, "ascpu", 50);
	safecopy(display_name, "", 50);
	safecopy(mainGeometry, "", 50);
	safecopy(state.bgcolor, "#303030", 50);
	safecopy(state.fgcolor, "#20b2aa", 50);

	safecopy(sys_color, "", 50);
	safecopy(nice_color, "", 50);
	safecopy(user_color, "", 50);
	safecopy(idle_color, "", 50);
}

/* print the usage for the tool */
void usage() 
{
        printf("Usage : ascpu [options ...]\n\n");
        printf("-V              print version and exit\n");
        printf("-h -H -help     print this message\n");
	printf("-title <name>	set the window/icon title to this name\n");
        printf("-u <secs>       the update interval in seconds\n");
        printf("-cpu <num>      the CPU number to show\n");
        printf("-samples <num>  the number of samples in the average\n");
        printf("-history <num> 	the number of samples for the running history\n");
        printf("-nonice		show nice CPU time as idle time\n");
        printf("-display <name> the name of the display to use\n");
        printf("-position <xy>  position on the screen (geometry)\n");
        printf("-withdrawn      start in withdrawn shape (for WindowMaker)\n");
        printf("-iconic         start iconized\n");
        printf("-standout       standing out rather than being pushed in\n");
#ifdef FOR_CLICK
	printf("-exe <command>	execute command on mouse click\n");
#endif
	printf("-dev <device>	use the specified file as stat device\n\n");
	printf("-bg <color>	background color\n");
	printf("-fg <color>	base foreground color\n");
	printf("-sys <color>	color for system CPU time\n");
	printf("-nice <color>	color for nice CPU time\n");
	printf("-user <color>	color for user CPU time\n");
	printf("-idle <color>	color for idle CPU time\n");
        printf("\n");
        exit(0);
}       

/* print the version of the tool */
void version()
{
        printf("ascpu : AfterStep CPU usage statistics monitor version 1.11\n");
}               

void    parsecmdline(int argc, char *argv[])
{
        char    *argument;
        int     i;

	/* parse the command line */
        for (i=1; i<argc; i++) {
                argument=argv[i];
                if (argument[0]=='-') {
                        if (!strncmp(argument,"-withdrawn",10)) {
                                withdrawn=1;
                        } else if (!strncmp(argument,"-iconic",7)) {
                                iconic=1;
                        } else if (!strncmp(argument,"-standout",9)) {
                                pushed_in=0;
                        } else if (!strncmp(argument,"-nonice",7)) {
                                state.no_nice=1;
                        } else if (!strncmp(argument,"-cpu",4)) {
                                if (++i >= argc)
                                        usage();
                                state.cpu_number = atoi(argv[i]);
                                if ( state.cpu_number < 0 )
                                        state.cpu_number = -1;
				if ( state.cpu_number >= MAX_CPU ) {
					state.cpu_number = -1;
					printf("ascpu: The maximum number of CPUs supported was restricted to %d at compilation\n", MAX_CPU);
				}
                        } else if (!strncmp(argument,"-samples",8)) {
                                if (++i >= argc)
                                        usage();
                                state.avg_samples = atoi(argv[i]);
                                if ( state.avg_samples < 1 )  
                                        state.avg_samples = AVG_SAMPLES;
#ifdef FOR_CLICK
                        } else if (!strncmp(argument,"-exe",4)) {
                                if (++i >= argc)
                                        usage();
                                safecopy(Command, argv[i], 252);
				strcat(Command," &");
#endif
                        } else if (!strncmp(argument,"-history",8)) {
                                if (++i >= argc)
                                        usage();
                                state.hist_samples = atoi(argv[i]);
                                if ( state.hist_samples < 1 )  
                                        state.hist_samples = HIST_SAMPLES;
                        } else if (!strncmp(argument,"-position",9)) {
                                if (++i >= argc)
                                        usage();
                                safecopy(mainGeometry, argv[i], 50);
                        } else if (!strncmp(argument,"-title",6)) {
                                if (++i >= argc)
                                        usage();
                                safecopy(window_name, argv[i], 50);
                        } else if (!strncmp(argument,"-display",8)) {
                                if (++i >= argc)
                                        usage();
                                safecopy(display_name, argv[i], 50);
                        } else if (!strncmp(argument,"-dev",4)) {
                                if (++i >= argc)
                                        usage();
                                safecopy(state.proc_stat_filename,argv[i],256);
			} else if (!strncmp(argument,"-bg",3)) {
				if (++i >= argc)
					usage();
				safecopy(state.bgcolor, argv[i], 50);
			} else if (!strncmp(argument,"-fg",3)) {
				if (++i >= argc)
					usage();
				safecopy(state.fgcolor, argv[i], 50);
			} else if (!strncmp(argument,"-sys",4)) {
				if (++i >= argc)
					usage();
				safecopy(sys_color, argv[i], 50);
				sys_color_defined = 1;
			} else if (!strncmp(argument,"-nice",5)) {
				if (++i >= argc)
					usage();
				safecopy(nice_color, argv[i], 50);
				nice_color_defined = 1;
			} else if (!strncmp(argument,"-user",5)) {
				if (++i >= argc)
					usage();
				safecopy(user_color, argv[i], 50);
				user_color_defined = 1;
			} else if (!strncmp(argument,"-idle",5)) {
				if (++i >= argc)
					usage();
				safecopy(idle_color, argv[i], 50);
				idle_color_defined = 1;
                        } else if (!strncmp(argument,"-u",2)) {
                                if (++i >= argc)
                                        usage();
                                state.update_interval = atoi(argv[i]);
                                if ( state.update_interval < 1 )  
                                        state.update_interval = CHK_INTERVAL;
                        } else if (!strncmp(argument,"-V",2)) { 
                                version();
                                exit(0);
                        } else if (!strncmp(argument,"-H",2)) {
                                version();
                                usage();
                        } else if (!strncmp(argument,"-h",2)) {
                                version();
                                usage();
                        } else {
                                version();
                                usage();
                        }       
                } else {
                        version();
                        usage();
                }       
        }       
 
}

int main(int argc, char** argv)
{
	defaults();
	parsecmdline(argc, argv);
	ascpu_initialize(argc, argv, 
			window_name,
			display_name, 
			mainGeometry, 
			withdrawn, 
			iconic, 
			pushed_in,
			sys_color_defined,
			sys_color,
			nice_color_defined,
			nice_color,
			user_color_defined,
			user_color,
			idle_color_defined,
			idle_color);
	while (1) {
		ascpu_update();
		usleep(X11_INTERVAL);
	}
	return(1);
}

